src/v/cluster/metrics_reporter.cc

```diff
@@ -0,0 +1,386 @@
+    boost::random::mt19937 mersenne_twister;
+    boost::random::seed_seq seed(hash.begin(), hash.end());
+    mersenne_twister.seed(seed);
+
+    boost::uuids::random_generator_mt19937 uuid_gen(mersenne_twister);
+
+    _cluster_uuid = fmt::format("{}", uuid_gen());
+}
+
+iobuf serialize_metrics_snapshot(
+  const metrics_reporter::metrics_snapshot& snapshot) {
+    rapidjson::StringBuffer sb;
+    rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+
+    json::rjson_serialize(writer, snapshot);
+    iobuf out;
+    out.append(sb.GetString(), sb.GetSize());
+
+    return out;
+}
+ss::future<http::client> metrics_reporter::make_http_client() {
+    rpc::base_transport::configuration client_configuration;
+
+    client_configuration.server_addr = unresolved_address(
+      ss::sstring(host), port);
+
+    client_configuration.disable_metrics = rpc::metrics_disabled::yes;
+
+    if (protocol == "https") {
+        client_configuration.credentials
+          = ss::make_shared<ss::tls::certificate_credentials>();
+        co_await client_configuration.credentials->set_system_trust();
+    }
+
+    co_return http::client(client_configuration, _as.local());
+}
+
+ss::future<> metrics_reporter::do_report_metrics() {
+    // skip reporting if current node is not raft0 leader, or we need to wait
+    // for next report
+    if (
+      !_raft0->is_leader()
+      || _last_success
+           > ss::lowres_clock::now()
+               - config::shard_local_cfg().metrics_reporter_report_interval()) {
+        co_return;
+    }
+
+    // try initializing cluster info, if it is already present this operation
+    // does nothig
+    co_await try_initialize_cluster_info();
+
+    // if not initialized, wait until next tick
+    if (_cluster_uuid == "") {
+        co_return;
+    }
+
+    // collect metrics
+    auto snapshot = co_await build_metrics_snapshot();
+    if (!snapshot) {
+        vlog(
+          _logger.trace,
+          "error collecting cluster metrics snapshot - {}",
+          snapshot.error().message());
+        co_return;
+    }
+    auto out = serialize_metrics_snapshot(snapshot.value());
+    auto header = make_header(out);
+    auto body = make_iobuf_input_stream(std::move(out));
+    try {
+        // prepare http client
+        auto client = co_await make_http_client();
+        auto timeout
+          = config::shard_local_cfg().metrics_reporter_tick_interval();
+        auto res = co_await client.get_connected(timeout, _logger);
+        // skip sending metrics, unable to connect
+        if (res != http::reconnect_result_t::connected) {
+            vlog(
+              _logger.trace,
+              "unable to send metrics report, connection timeout");
+            co_return;
+        }
+        auto resp_stream = co_await client.request(
+          std::move(header), body, timeout);
+        co_await resp_stream->prefetch_headers();
+        co_await resp_stream->shutdown();
+        _last_success = ss::lowres_clock::now();
```

Comment:
what is the behavior of the reporter if there is a error encountered while reporting here? presumably we report every 24 hours, but every 24 hours we also repeat trying if there is a problem. is the backoff 1 min?

Reply:
exactly we will try at worst every minute.